pkg/scheduler/framework/plugins/podtopologyspread/scoring.go

```diff
@@ -0,0 +1,255 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package podtopologyspread
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"sync/atomic"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+	"k8s.io/kubernetes/pkg/scheduler/algorithm/predicates"
+	framework "k8s.io/kubernetes/pkg/scheduler/framework/v1alpha1"
+)
+
+const postFilterStateKey = "PostFilter" + Name
+
+// postFilterState computed at PostFilter and used at Score.
+type postFilterState struct {
+	constraints []topologySpreadConstraint
+	// nodeNameSet is a string set holding all node names which have all constraints[*].topologyKey present.
+	nodeNameSet map[string]struct{}
+	// topologyPairToPodCounts is keyed with topologyPair, and valued with the number of matching pods.
+	topologyPairToPodCounts map[topologyPair]*int64
+}
+
+// Clone implements the mandatory Clone interface. We don't really copy the data since
+// there is no need for that.
+func (s *postFilterState) Clone() framework.StateData {
+	return s
+}
+
+// initialize iterates "filteredNodes" to filter out the nodes which don't have required topologyKey(s),
+// and initialize two maps:
+// 1) s.topologyPairToPodCounts: keyed with both eligible topology pair and node names.
+// 2) s.nodeNameSet: keyed with node name, and valued with a *int64 pointer for eligible node only.
+func (s *postFilterState) initialize(pod *v1.Pod, filteredNodes []*v1.Node) error {
+	constraints, err := filterTopologySpreadConstraints(pod.Spec.TopologySpreadConstraints, v1.ScheduleAnyway)
+	if err != nil {
+		return err
+	}
+	if constraints == nil {
+		return nil
+	}
+	s.constraints = constraints
+	for _, node := range filteredNodes {
+		if !nodeLabelsMatchSpreadConstraints(node.Labels, s.constraints) {
+			continue
+		}
+		for _, constraint := range s.constraints {
+			pair := topologyPair{key: constraint.topologyKey, value: node.Labels[constraint.topologyKey]}
+			if s.topologyPairToPodCounts[pair] == nil {
+				s.topologyPairToPodCounts[pair] = new(int64)
+			}
+		}
+		s.nodeNameSet[node.Name] = struct{}{}
```

Comment:
```s.nodeNameSet.Insert(node.Name)```

Reply:
I feel that we don't really need ```nodeNameSet```, if the comments below regarding the scores are true (lines 198 and 228), then we can just do ```nodeLabelsMatchSpreadConstraints(node.Labels, s.constraints)``` check at line 166 directly.

Reply:
> I feel that we don't really need `nodeNameSet`

It's because we need to distinguish with a **zero** value which is manually set, which means it has 0 matching pods; versus a **nil** value which is intentionally left there (explained in L74~L75, which means that node isn't qualified and we shouldn't treat it as 0, esp. in Score phase.